File: game/tv_show/tv_show_resources.py
import pygame
from pyvidplayer2 import Video

from config import Config


class TvShowResources:
    instructions = None
    videos_attract = {}
    videos_initial = {}
    videos_press_5 = {}
    videos_going_scylla = {}
    videos_ending = {}
    videos_have_luck = {}
    audio_attract = {}
    audio_initial = {}
    audio_press_5 = {}
    audio_going_scylla = {}
    audio_ending = {}
    audio_have_luck = {}
    press_5 = None

    @staticmethod
    def init():
        TvShowResources.press_5 = pygame.image.load("resources/images/press_5.png").convert_alpha()
        TvShowResources.instructions = {game_name: pygame.image.load("resources/images/instruction_" + game_name + ".png").convert() for
                    game_name in Config.GAMES.keys()}

        for country in Config.COUNTRIES:
            prefix = f"resources/videos/{country}/"
            audio_prefix = f"audio_for_videos/{country}/"
            
            TvShowResources.videos_attract[country] = Video(prefix + "attract_demo.avi", no_audio=True)
            TvShowResources.videos_initial[country] = Video(prefix + "hello_hello.avi", no_audio=True)
            TvShowResources.videos_press_5[country] = Video(prefix + "press_5.avi", no_audio=True)
            TvShowResources.videos_going_scylla[country] = Video(prefix + "scylla_cave.avi", no_audio=True)
            TvShowResources.videos_ending[country] = Video(prefix + "you_lost.avi", no_audio=True)
            TvShowResources.videos_have_luck[country] = Video(prefix + "have_luck.avi", no_audio=True)
            
            TvShowResources.audio_attract[country] = audio_prefix + "attract_demo.aac"
            TvShowResources.audio_initial[country] = audio_prefix + "hello_hello.aac"
            TvShowResources.audio_press_5[country] = audio_prefix + "press_5.aac"
            TvShowResources.audio_going_scylla[country] = audio_prefix + "scylla_cave.aac"
            TvShowResources.audio_ending[country] = audio_prefix + "you_lost.aac"
            TvShowResources.audio_have_luck[country] = audio_prefix + "have_luck.aac"

File: game-web/src/tvShow/tvShowResources.ts
/** Port of game/tv_show/tv_show_resources.py */

import { loadSurfaceRes } from '../resource';
import { Config } from '../config';

export class VideoWrapper {
  readonly element: HTMLVideoElement;

  constructor(src: string) {
    this.element = document.createElement('video');
    this.element.src = src;
    this.element.muted = true;   // audio handled separately via Web Audio
    this.element.preload = 'auto';
  }

  /** True while the video is playing (not paused / ended). */
  get active(): boolean {
    return !this.element.paused && !this.element.ended;
  }

  /** Restart from the beginning and start playing. */
  restart(): void {
    this.element.currentTime = 0;
    void this.element.play();
  }

  /** Stop playback and rewind. */
  stop(): void {
    this.element.pause();
    this.element.currentTime = 0;
  }

  /** Draw the current video frame to a 2D context at (x, y). */
  draw(ctx: OffscreenCanvasRenderingContext2D, x: number, y: number): void {
    ctx.drawImage(this.element, x, y);
  }
}

export class TvShowResources {
  static press5: ImageBitmap | null = null;
  static instructions: Record<string, ImageBitmap> = {};

  static videosAttract:     Record<string, VideoWrapper> = {};
  static videosInitial:     Record<string, VideoWrapper> = {};
  static videosPress5:      Record<string, VideoWrapper> = {};
  static videosGoingScylla: Record<string, VideoWrapper> = {};
  static videosEnding:      Record<string, VideoWrapper> = {};
  static videosHaveLuck:    Record<string, VideoWrapper> = {};

  static audioAttract:     Record<string, string> = {};
  static audioInitial:     Record<string, string> = {};
  static audioPress5:      Record<string, string> = {};
  static audioGoingScylla: Record<string, string> = {};
  static audioEnding:      Record<string, string> = {};
  static audioHaveLuck:    Record<string, string> = {};

  static async init(): Promise<void> {
    TvShowResources.press5 = await loadSurfaceRes('images/press_5.png');

    const instructionPromises = Object.keys(Config.GAMES).map(async (gameName) => {
      const bitmap = await loadSurfaceRes(`images/instruction_${gameName}.png`);
      TvShowResources.instructions[gameName] = bitmap;
    });
    await Promise.all(instructionPromises);

    for (const country of Config.COUNTRIES) {
      const videoPrefix = `${import.meta.env.BASE_URL}videos/${country}/`;
      const audioPrefix = `audio_for_videos/${country}/`;

      TvShowResources.videosAttract[country]     = new VideoWrapper(videoPrefix + 'attract_demo.webm');
      TvShowResources.videosInitial[country]     = new VideoWrapper(videoPrefix + 'hello_hello.webm');
      TvShowResources.videosPress5[country]      = new VideoWrapper(videoPrefix + 'press_5.webm');
      TvShowResources.videosGoingScylla[country] = new VideoWrapper(videoPrefix + 'scylla_cave.webm');
      TvShowResources.videosEnding[country]      = new VideoWrapper(videoPrefix + 'you_lost.webm');
      TvShowResources.videosHaveLuck[country]    = new VideoWrapper(videoPrefix + 'have_luck.webm');

      TvShowResources.audioAttract[country]     = audioPrefix + 'attract_demo.wav';
      TvShowResources.audioInitial[country]     = audioPrefix + 'hello_hello.wav';
      TvShowResources.audioPress5[country]      = audioPrefix + 'press_5.wav';
      TvShowResources.audioGoingScylla[country] = audioPrefix + 'scylla_cave.wav';
      TvShowResources.audioEnding[country]      = audioPrefix + 'you_lost.wav';
      TvShowResources.audioHaveLuck[country]    = audioPrefix + 'have_luck.wav';
    }
  }
}
